RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,566 @@
+
+  Support source types:
+  * Filesystem based repository
+
+    A directory on disk can be specified as a Template Repository.
+
+  * Git based repository
+
+    A git repository (For example Github, GitLab or in internal git repo)
+
+  * `default`
+
+    This special source type will use whatever the Template Engine considers the default.  In the case of the PDK, this would be the cached `pdk-templates` repo that is packaged as part of the PDK. Other tools using the Template Engine may specify a different location.
+
+  If no template sources are specified `default` is used.
+
+* `templates` element (Array of string). List of templates (The template directory name) to be applied **IN ORDER** (Order is important.  See [Resolution Order](#resolution_order))
+
+Example V1 `.sync.yml`
+
+``` yaml
+---
+Gemfile:
+  # Gemfile template customisations
+
+.rubocop.yml:
+  # Rubocop template customisations
+
+# ... etc
+```
+
+Example `.sync.yml` with V2 settings
+
+``` yaml
+---
+pdk_template:
+  version: 2
+  template_sources:
+    -
+      type: git
+      location: 'https://github.com/user/custom-templates.git'
+    - default
+  templates:
+    - appveyor_ci
+    - travis_ci
+    - rubocop
+    - litmus_tests
+
+gemfile:
+  # Gemfile template customisations
+
+rubocop:
+  # Rubocop template customisations
+
+# ... etc
+```
+
+#### Filesystem based template sources
+
+Example:
+
+``` yaml
+---
+pdk_template:
+  template_sources:
+    -
+      type: filesystem
+      location: 'C:/templates/custom'
+    #  Unix style path
+    # -
+    #   type: filesystem
+    #   location: '/templates/custom'
+```
+
+| YAML Element | Description |
+| ------------ | ----------- |
+| type         | Must be `filesystem` |
+| location     | The path the repository directory |
+
+#### Git based template sources
+
+``` yaml
+---
+pdk_template:
+  template_sources:
+    -
+      type: git
+      location: 'https://github.com/user/custom-templates.git'
+      ref: master
+```
+
+| YAML Element | Description |
+| ------------ | ----------- |
+| type         | Must be `git` |
+| location     | The git remote to clone |
+| ref          | (Optional) The git reference to checkout. Default is what the git remote deems the default branch |
+
+### Partially overriding default templates
+
+> TODO. Should this be possible or encouraged?
+>
+> The idea here is to ONLY modify the ERB files so people can "tweak" default templates. This will encourage people to use the default templates more, instead of forking.
+
+### Resolution order
+
+The resolution order used in the V2 templates is the top-most item in the list will be used.
+
+> TODO: What happens with conflicting settings files.  Order is important, first one wins.
+> For `always_apply` modules they are last, but in alphabetical order (Needs some kind of deterministic order)
```

Comment:
Making the template list/application order important seems like a source of hard to understand/debug behavior.

I understand the need for there to be a pre-determined hierarchy for default versus per-module settings and such, but is there a use case where multiple templates are trying to publish the same shared setting (with different values) and this shouldn't just be an error? And non-exported settings should be internal to each template so application order shouldn't matter at that point.